### PR TITLE
131 resolve-reloading-issue-on-pinned-courses-view

### LINKED ISSUE #131: Resolve Reloading Issue on Pinned Courses View
## Objective

Users should be able to reload the page on the Pinned Courses view regardless of whether there are pinned courses or not.
##

This is how it should look like when refreshing the view:
Screenshot_20231214_043714-1

_Originally posted by @carlobortolan in https://github.com/TUM-Dev/gocast_mobile/pull/125#pullrequestreview-1780895536_

### PR DESCRIPTION
**Pull request by cbetul**
### Add

Users can be able to **reload** the page on the `Pinned Courses View` regardless of whether there are pinned courses or not.
##
### Resolves
closes #131

### CODE REVIEW

**Review by carlobortolan — CHANGES_REQUESTED** [review]
When scrolling up/down on the pinned course page, the courses disappear; please fix this bug before merging into dev:

https://github.com/TUM-Dev/gocast_mobile/assets/106114526/a8b6ce85-b3cb-42fd-a083-c7ff45100d01

This is the console log:
```shell
I/flutter (10827): ┌───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
I/flutter (10827): │ #0   PinnedHandler.fetchUserPinned (package:gocast_mobile/base/networking/api/handler/pinned_handler.dart:20:13)
I/flutter (10827): │ #1   UserViewModel.fetchUserPinned (package:gocast_mobile/view_models/user_view_model.dart:85:55)
I/flutter (10827): ├┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄
I/flutter (10827): │ 💡 Fetching user pinned
I/flutter (10827): └───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
I/flutter (10827): ┌───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
I/flutter (10827): │ #0   GrpcHandler.callGrpcMethod (package:gocast_mobile/base/networking/api/handler/grpc_handler.dart:45:13)
I/flutter (10827): │ #1   PinnedHandler.fetchUserPinned (package:gocast_mobile/base/networking/api/handler/pinned_handler.dart:21:25)
I/flutter (10827): ├┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄
I/flutter (10827): │ 🐛 callGrpcMethod: Initiating gRPC call
I/flutter (10827): └───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
I/flutter (10827): ┌───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
I/flutter (10827): │ #0   TokenHandler.loadToken (package:gocast_mobile/base/networking/api/handler/token_handler.dart:50:15)
I/flutter (10827): │ #1   <asynchronous suspension>
I/flutter (10827): ├┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄
I/flutter (10827): │ 💡 Token successfully loaded for key: jwt
I/flutter (10827): └───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
I/flutter (10827): ┌───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
I/flutter (10827): │ #0   PinnedHandler.fetchUserPinned.<anonymous closure> (package:gocast_mobile/base/networking/api/handler/pinned_handler.dart:24:17)
I/flutter (10827): │ #1   <asynchronous suspension>
I/flutter (10827): ├┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄
I/flutter (10827): │ 💡 User pinned fetched successfully
I/flutter (10827): └───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
I/flutter (10827): ┌───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
I/flutter (10827): │ #0   PinnedHandler.fetchUserPinned.<anonymous closure> (package:gocast_mobile/base/networking/api/handler/pinned_handler.dart:25:17)
I/flutter (10827): │ #1   <asynchronous suspension>
I/flutter (10827): ├┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄
I/flutter (10827): │ 🐛 User pinned: [id: 1
I/flutter (10827): │ 🐛 name: Einführung Brauereiwesen
I/flutter (10827): │ 🐛 slug: brauereiwesen
I/flutter (10827): │ 🐛 semester: {
I/flutter (10827): │ 🐛   teachingTerm: S
I/flutter (10827): │ 🐛   year: 2022
I/flutter (10827): │ 🐛 }
I/flutter (10827): │ 🐛 VODEnabled: true
I/flutter (10827): │ 🐛 downloadsEnabled: true
I/flutter (10827): │ 🐛 anonymousChatEnabled: true
I/flutter (10827): │ 🐛 , id: 2
I/flutter (10827): │ 🐛 name: Spieleentwicklung für Dummies
I/flutter (10827): │ 🐛 slug: games101
I/flutter (10827): │ 🐛 semester: {
I/flutter (10827): │ 🐛   teachingTerm: S
I/flutter (10827): │ 🐛   year: 2022
I/flutter (10827): │ 🐛 }
I/flutter (10827): │ 🐛 VODEnabled: true
I/flutter (10827): │ 🐛 downloadsEnabled: true
I/flutter (10827): │ 🐛 anonymousChatEnabled: true
I/flutter (10827): │ 🐛 , id: 3
I/flutter (10827): │ 🐛 name: Praktikum: Golang
I/flutter (10827): │ 🐛 slug: godev
I/flutter (10827): │ 🐛 semester: {
I/flutter (10827): │ 🐛   teachingTerm: W
I/flutter (10827): │ 🐛   year: 2021
I/flutter (10827): │ 🐛 }
I/flutter (10827): │ 🐛 VODEnabled: true
I/flutter (10827): │ 🐛 downloadsEnabled: true
I/flutter (10827): │ 🐛 anonymousChatEnabled: true
I/flutter (10827): │ 🐛 , id: 5
I/flutter (10827): │ 🐛 name: New Course
I/flutter (10827): │ 🐛 slug: new-course
I/flutter (10827): │ 🐛 semester: {
I/flutter (10827): │ 🐛   teachingTerm: W
I/flutter (10827): │ 🐛   year: 2023
I/flutter (10827): │ 🐛 }
I/flutter (10827): │ 🐛 VODEnabled: true
I/flutter (10827): │ 🐛 downloadsEnabled: true
I/flutter (10827): │ 🐛 anonymousChatEnabled: true
I/flutter (10827): │ 🐛 ]
I/flutter (10827): └───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
D/EGL_emulation(14796): app_time_stats: avg=33.71ms min=9.74ms max=339.35ms count=31
Another exception was thrown: Vertical viewport was given unbounded height.
Another exception was thrown: RenderBox was not laid out: RenderViewport#62b44 NEEDS-LAYOUT NEEDS-PAINT
NEEDS-COMPOSITING-BITS-UPDATE
Another exception was thrown: RenderBox was not laid out: RenderViewport#62b44 NEEDS-PAINT
NEEDS-COMPOSITING-BITS-UPDATE
Another exception was thrown: RenderBox was not laid out: RenderIgnorePointer#0cf56 relayoutBoundary=up13
NEEDS-PAINT NEEDS-COMPOSITING-BITS-UPDATE
Another exception was thrown: RenderBox was not laid out: RenderSemanticsAnnotations#20235 relayoutBoundary=up12
NEEDS-PAINT NEEDS-COMPOSITING-BITS-UPDATE
Another exception was thrown: RenderBox was not laid out: RenderPointerListener#8c8ce relayoutBoundary=up11
NEEDS-PAINT NEEDS-COMPOSITING-BITS-UPDATE
Another exception was thrown: RenderBox was not laid out: RenderSemanticsGestureHandler#1828c relayoutBoundary=up10
NEEDS-PAINT NEEDS-COMPOSITING-BITS-UPDATE
Another exception was thrown: RenderBox was not laid out: RenderPointerListener#55b7a relayoutBoundary=up9
NEEDS-PAINT NEEDS-COMPOSITING-BITS-UPDATE
Another exception was thrown: RenderBox was not laid out: _RenderScrollSemantics#da5d6 relayoutBoundary=up8
NEEDS-PAINT NEEDS-COMPOSITING-BITS-UPDATE
Another exception was thrown: RenderBox was not laid out: RenderTransform#efa0b relayoutBoundary=up7 NEEDS-PAINT
NEEDS-COMPOSITING-BITS-UPDATE
Another exception was thrown: RenderBox was not laid out: RenderClipRect#8562c relayoutBoundary=up6 NEEDS-PAINT
NEEDS-COMPOSITING-BITS-UPDATE
Another exception was thrown: RenderBox was not laid out: RenderStack#026a2 relayoutBoundary=up5 NEEDS-PAINT
NEEDS-COMPOSITING-BITS-UPDATE
Another exception was thrown: RenderBox was not laid out: RenderRepaintBoundary#e0b8e relayoutBoundary=up4
NEEDS-PAINT NEEDS-COMPOSITING-BITS-UPDATE
Another exception was thrown: 'package:flutter/src/rendering/sliver_multi_box_adaptor.dart': Failed assertion: line
544 pos 12: 'child.hasSize': is not true.
Another exception was thrown: Null check operator used on a null value

...
```

File: lib/views/course_view/components/pinned_course_list.dart
Comment (GravityDarkLab) [inline lib/views/course_view/components/pinned_course_list.dart:1]: Nested `ListView.builder` can cause performance issues.

**Review by GravityDarkLab — CHANGES_REQUESTED** [review]

File: lib/views/course_view/components/pinned_course_list.dart
Comment (cbetul) [reply lib/views/course_view/components/pinned_course_list.dart:1]: fixed

File: lib/views/course_view/pinned_courses_view/pinned_courses_content_view.dart [hunk 1]
```diff
@@ -16,18 +16,21 @@ import 'package:gocast_mobile/views/course_view/components/pinned_course_card.da
 class PinnedCoursesContentView extends ConsumerWidget {
   final String title;
   final List<PinnedCourseCard> pinnedCourseCards;
+  final Future<void> Function()? onRefresh;
```
Comment (carlobortolan) [inline lib/views/course_view/pinned_courses_view/pinned_courses_content_view.dart:19]: This `onRefresh` attribute is not needed and can be removed.

File: lib/views/course_view/pinned_courses_view/pinned_courses_content_view.dart [hunk 2]
```diff
@@ -16,18 +16,21 @@ import 'package:gocast_mobile/views/course_view/components/pinned_course_card.da
...
 
   const PinnedCoursesContentView({
     super.key,
     required this.title,
     required this.pinnedCourseCards,
+    this.onRefresh,
```
Comment (carlobortolan) [inline lib/views/course_view/pinned_courses_view/pinned_courses_content_view.dart:25]: Dead code (See comment above)

File: lib/views/course_view/pinned_courses_view/pinned_courses_content_view.dart [hunk 3]
```diff
@@ -16,18 +16,21 @@ import 'package:gocast_mobile/views/course_view/components/pinned_course_card.da
...
   });
 
   @override
   Widget build(BuildContext context, WidgetRef ref) {
     return PinnedCourseList(
       title: title,
       pinnedCoursesCard: pinnedCourseCards,
+      onRefresh: onRefresh,
```
Comment (carlobortolan) [inline lib/views/course_view/pinned_courses_view/pinned_courses_content_view.dart:33]: Dead code (See comment above)

File: lib/views/course_view/components/pinned_course_list.dart [hunk 4]
```diff
@@ -41,19 +41,40 @@ class PinnedCourseList extends ConsumerWidget {
...
-          : ListView.builder(
-              itemCount: pinnedCoursesCard.length,
-              itemBuilder: (BuildContext context, int index) {
+      child: NotificationListener(
+        onNotification: (ScrollNotification scrollInfo) {
+          if (scrollInfo is ScrollEndNotification && scrollInfo.metrics.extentBefore == 0.0) {
+            onRefresh?.call();
+          }
+          return false;
+        },
```
Comment (carlobortolan) [inline lib/views/course_view/components/pinned_course_list.dart:50]: Dead code (See comment above)

**Review by carlobortolan — CHANGES_REQUESTED** [review]
The course-disappearing bug is fixed and the code works fine now, just make sure to remove unused code (in this case `Future<void> Function()? onRefresh`).
Once that's fixed, you can merge this PR :+1:

File: lib/views/course_view/pinned_courses_view/pinned_courses_content_view.dart (hunk 1 above)
Comment (cbetul) [reply lib/views/course_view/pinned_courses_view/pinned_courses_content_view.dart]: fixed

File: lib/views/course_view/pinned_courses_view/pinned_courses_content_view.dart (hunk 3 above)
Comment (cbetul) [reply lib/views/course_view/pinned_courses_view/pinned_courses_content_view.dart]: removed

File: lib/views/course_view/components/pinned_course_list.dart (hunk 4 above)
Comment (cbetul) [reply lib/views/course_view/components/pinned_course_list.dart]: removed

File: lib/views/course_view/pinned_courses_view/pinned_courses_content_view.dart (hunk 2 above)
Comment (cbetul) [reply lib/views/course_view/pinned_courses_view/pinned_courses_content_view.dart]: fixed

**Review by carlobortolan — CHANGES_REQUESTED** [review]
There still is dead code in the `pinned_course_list.dart`, lines 17 and 23:
https://github.com/TUM-Dev/gocast_mobile/blob/7e03de7cc6087cb822ea643f6e5300827d084b0e/lib/views/course_view/components/pinned_course_list.dart#L17C1-L17C44

**Review by carlobortolan — APPROVED** [review]

**Review by GravityDarkLab — APPROVED** [review]

**Pull request merged**